projects/pgai/pgai/vectorizer/embedders/litellm.py

```diff
@@ -0,0 +1,132 @@
+            implementation.
+        model (str): The name of the embedding model.
+        api_key_name (str): The API key name.
+        extra_options (dict): Additional litellm-specific options
+    """
+
+    implementation: Literal["litellm"]
+    model: str
+    api_key_name: str | None = None
+    extra_options: dict[str, Any] = {}
+
+    @override
+    async def embed(self, documents: list[str]) -> Sequence[EmbeddingVector]:
+        """
+        Embeds a list of documents into vectors using LiteLLM.
+
+        Args:
+            documents (list[str]): A list of documents to be embedded.
+
+        Returns:
+            Sequence[EmbeddingVector | ChunkEmbeddingError]: The embeddings or
+            errors for each document.
+        """
+        await logger.adebug(f"Chunks produced: {len(documents)}")
+        return await self._batcher.batch_chunks_and_embed(documents)
+
+    @cached_property
+    def _batcher(self) -> BatchApiCaller[StringDocument]:
+        return BatchApiCaller(self._max_chunks_per_batch(), self.call_embed_api)
+
+    @override
+    def _max_chunks_per_batch(self) -> int:
+        print(f"model: {self.model}")
+        _, custom_llm_provider, _, _ = litellm.get_llm_provider(self.model)  # type: ignore
+        match custom_llm_provider:
+            case "cohere":
+                return 96  # see https://docs.cohere.com/v1/reference/embed#request.body.texts
+            case "openai":
+                return 2048  # see https://platform.openai.com/docs/api-reference/embeddings/create
+            case "azure":
+                return 2048  # https://learn.microsoft.com/en-us/azure/ai-services/openai/how-to/embeddings?tabs=console#verify-inputs-dont-exceed-the-maximum-length
+            case "bedrock":
+                return 96  # NOTE: currently (Jan 2025) Bedrock only supports embeddings with Cohere or Titan models. The Titan API only processes one input per request, which LiteLLM already handles under the hood. We assume that the Cohere API has the same input limits as above.
+            case "huggingface":
+                return 2048  # NOTE: There is not documented limit. In testing we got a response for a request with 10k (short) inputs.
+            case "mistral":
+                return 128  # FIXME: this is chosen somewhat arbitrarily. There is no limit on # chunks per batch, but there is a (low) limit on # tokens per batch (16384)
+            case "vertex":
+                return 250  # FIXME: this applies to `us-central-1` only (otherwise 5). Additionally there is a limit on # tokens per batch (20k) see https://cloud.google.com/vertex-ai/generative-ai/docs/embeddings/get-text-embeddings#get_text_embeddings_for_a_snippet_of_text
+            case "voyage":
+                return 128  # see https://docs.voyageai.com/reference/embeddings-api
+            case _:
+                raise UnknownProviderError(custom_llm_provider)
```

Comment:
props for actually figuring all these out ngl I would've set it just to some small number and called it a day.

Reply:
Thanks.